packages/astro/src/vite-plugin-markdown/index.ts

```diff
@@ -32,247 +23,95 @@ function safeMatter(source: string, id: string) {
-				return {
-					code: `
-						// Static
-						export const frontmatter = ${escapeViteEnvReferences(JSON.stringify(frontmatter))};
-						export const file = ${JSON.stringify(fileId)};
-						export const url = ${JSON.stringify(fileUrl)};
-						export function rawContent() {
-							return ${escapeViteEnvReferences(JSON.stringify(rawContent))};
-						}
-						export async function compiledContent() {
-							return load().then((m) => m.compiledContent());
-						}
-						export function $$loadMetadata() {
-							return load().then((m) => m.$$metadata);
-						}
-
-						// Deferred
-						export default async function load() {
-							return (await import(${JSON.stringify(fileId + MARKDOWN_CONTENT_FLAG)}));
-						}
-						export function Content(...args) {
-							return load().then((m) => m.default(...args));
-						}
-						Content.isAstroComponentFactory = true;
-						export function getHeadings() {
-							return load().then((m) => m.metadata.headings);
-						}
-						export function getHeaders() {
-							console.warn('getHeaders() have been deprecated. Use getHeadings() function instead.');
-							return load().then((m) => m.metadata.headings);
-						};`,
-					map: null,
-				};
-			}
-
-			// A markdown file is being rendered! This markdown file was either imported
-			// directly as a page in Vite, or it was a deferred render from a JS module.
-			// This returns the compiled markdown -> astro component that renders to HTML.
+		name: 'astro:markdown',
+		// Why not the "transform" hook instead of "load" + readFile?
+		// A: Vite transforms all "import.meta.env" references to their values before
+		// passing to the transform hook. This lets us get the truly raw value
+		// to escape "import.meta.env" ourselves.
+		async load(id) {
 			if (id.endsWith('.md')) {
-				const filename = normalizeFilename(id);
-				const source = await fs.promises.readFile(filename, 'utf8');
-				const renderOpts = config.markdown;
-				const isAstroFlavoredMd = config.legacy.astroFlavoredMarkdown;
-
-				const fileUrl = new URL(`file://${filename}`);
-
-				// Extract special frontmatter keys
-				let { data: frontmatter, content: markdownContent } = safeMatter(source, filename);
-
-				// Turn HTML comments into JS comments while preventing nested `*/` sequences
-				// from ending the JS comment by injecting a zero-width space
-				// Inside code blocks, this is removed during renderMarkdown by the remark-escape plugin.
-				if (isAstroFlavoredMd) {
-					markdownContent = markdownContent.replace(
-						/<\s*!--([^-->]*)(.*?)-->/gs,
-						(whole) => `{/*${whole.replace(/\*\//g, '*\u200b/')}*/}`
-					);
-				}
-
-				let renderResult = await renderMarkdown(markdownContent, {
-					...renderOpts,
-					fileURL: fileUrl,
-					isAstroFlavoredMd,
+				const { fileId, fileUrl } = getFileInfo(id, config);
+				const rawFile = String(await fs.promises.readFile(fileId));
```

Comment:
nit, but adding `, 'utf-8')` to this function call will give you a string.

Reply:
TIL. Nice!